compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -673,7 +673,41 @@ pub fn orphan_check(tcx: TyCtxt<'_>, impl_def_id: DefId) -> Result<(), OrphanChe
         return Ok(());
     }
 
-    orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |ty| Ok(ty)).unwrap()
+    orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |ty| {
+        let delay_bug = || {
+            tcx.sess.span_delayed_bug(
+                tcx.def_span(impl_def_id),
+                format!(
+                    "orphan check: failed to normalize `{trait_ref}` while checking {impl_def_id:?}"
+                ),
+            )
+        };
+
+        let infcx = tcx.infer_ctxt().intercrate(true).build();
+        let cause = ObligationCause::dummy();
+        let param_env = tcx.param_env(impl_def_id);
+
+        let ocx = ObligationCtxt::new(&infcx);
+        let ty = ocx.normalize(&cause, param_env, ty);
+        let ty = infcx.resolve_vars_if_possible(ty);
+        if !ocx.select_where_possible().is_empty() {
+            delay_bug();
+        }
+
+        if infcx.next_trait_solver()
+            && let ty::Alias(..) = ty.kind()
+        {
+            let mut fulfill_cx = <dyn TraitEngine<'_>>::new(&infcx);
+            infcx
+                .at(&cause, param_env)
+                .structurally_normalize(ty, &mut *fulfill_cx)
+                .map(|ty| infcx.resolve_vars_if_possible(ty))
+                .or_else(|_| Ok(Ty::new_error(tcx, delay_bug())))
+        } else {
+            Ok(ty)
+        }
```

Comment:
the ICEs happen here when trying to normalize, it uses `trait_ref_is_knowable` which asserts that we either have params or infer vars. I think we should:
- minimize one of the regressions and add it as a UI test
- instantiate generic params in `ty` with inference vars use an empty param env. Using a non-empty param env in the old solver causes issues anyways, as we incompletely prefer them over e.g. impls

Reply:
Right, and I also need to check why the IOOBs / escaping infer vars happen or if it's already fixed. Lastly, there are pass→error regressions all of which might reduce to `orphan-check-projection-does-cover.rs`, need to investigate.